File: MoonEngine-Root/ECS/src/main/java/com/greentree/engine/moon/ecs/ClassSetEntity.java
package com.greentree.engine.moon.ecs;

import java.io.Externalizable;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.util.Iterator;

import com.greentree.commons.action.observable.TypedObjectObservable;
import com.greentree.commons.util.iterator.IteratorUtil;
import com.greentree.engine.moon.ecs.ClassSet.LockClassSet;
import com.greentree.engine.moon.ecs.annotation.AnnotationUtil;
import com.greentree.engine.moon.ecs.component.Component;

public final class ClassSetEntity implements Entity, Externalizable, Cloneable {
	
	
	private final ClassSet<Component> components = new ComponentClassSet();
	
	public ClassSetEntity() {
	}
	
	@Override
	public void clear() {
		components.clear();
	}
	
	@Override
	public ClassSetEntity copy() {
		final var clone = new ClassSetEntity();
		copyTo(clone);
		return clone;
	}
	
	@Override
	public boolean contains(Class<? extends Component> componentClass) {
		return components.contains(componentClass);
	}
	
	@Override
	public boolean contains(Component component) {
		return components.contains(component);
	}
	
	@Override
	public <T extends Component> T get(Class<T> componentClass) {
		return components.get(componentClass);
	}
	
	@Override
	public TypedObjectObservable<? extends Class<? extends Component>, ? extends Component> getAddAction() {
		return components.getAddAction();
	}
	
	@Override
	public TypedObjectObservable<? extends Class<? extends Component>, ? extends Component> getRemoveAction() {
		return components.getRemoveAction();
	}
	
	@Override
	public boolean isEmpty() {
		return components.isEmpty();
	}
	
	@Override
	public Iterator<Component> iterator() {
		return components.iterator();
	}
	
	@Override
	public LockClassSet<Component> lock() {
		return components.lock();
	}
	
	@Override
	public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
		var size = in.readInt();
		try(final var lock = lock()) {
			while(size-- > 0) {
				final var c = (Component) in.readObject();
				lock.add(c);
			}
		}
	}
	
	@Override
	public int size() {
		return components.size();
	}
	
	
	@Override
	public String toString() {
		return "Entity " + IteratorUtil.toString(components);
	}
	
	@Override
	public void writeExternal(ObjectOutput out) throws IOException {
		out.writeInt(components.size());
		for(var c : this)
			out.writeObject(c);
	}
	
	
	public static final class ComponentClassSet extends ClassSet<Component> {
		
		private static final long serialVersionUID = 1L;
		
		public ComponentClassSet() {
		}
		
		@Override
		protected Iterable<? extends Class<? extends Component>> getClassRequired(
				Class<? extends Component> cls) {
			return AnnotationUtil.getRequiredComponent(cls);
		}
		
	}
	
}

File: MoonEngine-Root/ECS/src/main/java/com/greentree/engine/moon/ecs/ClassSetEntity.kt
package com.greentree.engine.moon.ecs

import com.greentree.commons.util.iterator.IteratorUtil
import com.greentree.engine.moon.ecs.component.Component
import com.greentree.engine.moon.kernel.AnnotationUtil
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.io.ObjectInput
import java.io.ObjectInputStream
import java.io.ObjectOutput
import java.util.stream.Stream

class ClassSetEntity : PrototypeEntity, Cloneable {

	private val components = ComponentClassSet()

	override fun clear() {
		components.clear()
	}

	override fun contains(componentClass: Class<out Component>): Boolean {
		return components.contains(componentClass)
	}

	override fun contains(component: Component): Boolean {
		return components.contains(component)
	}

	override fun <T : Component> get(componentClass: Class<T>): T {
		return components[componentClass]
	}

	override fun copy(): ClassSetEntity {
		val clone = ClassSetEntity()
		copyTo(clone)
		return clone
	}

	override fun isEmpty(): Boolean {
		return components.isEmpty
	}

	override fun iterator(): MutableIterator<Component> {
		return components.iterator()
	}

	override fun lock(): ComponentLock {
		return ClassSetComponentLock(components.lock())
	}

	override fun size(): Int {
		return components.size()
	}

	override fun toString(): String {
		return "Entity " + IteratorUtil.toString(components)
	}

	override fun save(output: ObjectOutput) {
		output.writeInt(components.size())
		for (c in components)
			output.writeObject(c)
	}

	class ComponentClassSet : ClassSet<Component>() {

		override fun getClassRequired(
			cls: Class<out Component>,
		): Iterable<Class<out Component>> {
			return getRequiredComponent(cls).toList()
		}

		companion object {

			private const val serialVersionUID = 1L

			fun getRequiredComponent(cls: Class<*>): Stream<Class<out Component>> {
				val requiredComponents = AnnotationUtil.getAnnotation(cls, RequiredComponent::class.java)
				return requiredComponents.stream().flatMap { Stream.of(*it.value) }.map { it.java }
			}
		}
	}

	companion object {

		@JvmOverloads
		@JvmStatic
		fun load(bytes: ByteArray, offset: Int = 0, length: Int = bytes.size): ClassSetEntity {
			return use(ByteArrayInputStream(bytes, offset, length)) {
				load(it)
			}
		}

		@JvmStatic
		fun load(input: InputStream): ClassSetEntity {
			if (input is ObjectInput)
				return load(input as ObjectInput)
			return load(ObjectInputStream(input) as ObjectInput)
		}

		@JvmStatic
		fun load(input: ObjectInput): ClassSetEntity {
			val result = ClassSetEntity()
			use(result.components.lock()) { lock ->
				repeat(input.readInt()) {
					lock.add(input.readObject() as Component)
				}
			}
			return result
		}
	}
}